edb/edgeql/parser/grammar/convert.py

```diff
@@ -0,0 +1,418 @@
+                isinstance(item, EBNF_Multiple) and
+                len(item.inner) == 1 and
+                is_single_reference(item.inner[0])
+            )
+    def get_single_reference_inner(item: EBNF_Item) -> EBNF_Item:
+        if isinstance(item, EBNF_Reference):
+            return item
+        elif isinstance(item, EBNF_Multiple):
+            return get_single_reference_inner(item.inner[0])
+        else:
+            raise NotImplementedError
+
+    single_references: Dict[str, EBNF_Item] = {
+        production.name: get_single_reference_inner(production.item)
+        for production in productions
+        if is_single_reference(production.item)
+    }
+
+    productions = [
+        production
+        for production in productions
+        if not is_single_reference(production.item)
+    ]
+
+    def replace_direct_references(item: EBNF_Item
+            ) -> Tuple[EBNF_Item, bool]:
+        return replace_references(item, single_references)
+
+    # inline single optionals
+
+    def is_single_optional(item: EBNF_Item) -> bool:
+        return isinstance(item, EBNF_Optional) or (
+                isinstance(item, EBNF_Multiple) and
+                len(item.inner) == 1 and
+                is_single_optional(item.inner[0])
+            )
+    def get_single_optional_inner(item: EBNF_Item) -> EBNF_Item:
+        if isinstance(item, EBNF_Optional):
+            return item.inner
+        elif isinstance(item, EBNF_Multiple):
+            return get_single_optional_inner(item.inner[0])
+        else:
+            raise NotImplementedError
+
+    single_optionals: Dict[str, EBNF_Item] = {
+        production.name: get_single_optional_inner(production.item)
+        for production in productions
+        if is_single_optional(production.item)
+    }
+
+    productions = [
+        production
+        for production in productions
+        if not is_single_optional(production.item)
+    ]
+
+    def replace_single_optional_references(item: EBNF_Item
+            ) -> Tuple[EBNF_Item, bool]:
+        return replace_references(item, single_optionals)
+
+    # replace choices of optionals with optional of choice
+
+    def replace_choice_of_options(item: EBNF_Item
+            ) -> Tuple[EBNF_Item, bool]:
+
+        if isinstance(item, EBNF_Choice):
+            inner_options = [
+                inner_item
+                for inner_item in item.inner
+                if isinstance(inner_item, EBNF_Optional)
+            ]
+
+            if len(inner_options) == len(item.inner):
+                return EBNF_Optional(EBNF_Choice([
+                    inner_option.inner
+                    for inner_option in inner_options
+                ])), True
+
+        return item, False
+
+    # apply replacements until no changes are made
+
+    def replace_repeatedly(
+                item: EBNF_Item,
+                funcs: List[Callable[[EBNF_Item], Tuple[EBNF_Item, bool]]]
+            ) -> EBNF_Item:
```

Comment:
We aren't totally consistent about these things, but typically edgedb style for this kind of definition is to have the closing paren lined up with the `def` and the body indented one step. (This is what black does, among others)